core/src/main/java/com/linecorp/armeria/server/logging/structured/StructuredLog.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.logging.structured;
+
+import java.util.concurrent.TimeUnit;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.logging.RequestLog;
+import com.linecorp.armeria.common.logging.ResponseLog;
+import com.linecorp.armeria.common.util.TextFormatter;
+
+/**
+ * A representation and constructor of a service log which holds only very common fields that are protocol
+ * agnostic.
+ */
+public abstract class StructuredLog {
+    private final long timestampMillis;
+    private final long responseTimeNanos;
+    private final long requestSize;
+    private final long responseSize;
+
+    StructuredLog(long timestampMillis, long responseTimeNanos, long requestSize, long responseSize) {
+        this.timestampMillis = timestampMillis;
+        this.responseTimeNanos = responseTimeNanos;
+        this.requestSize = requestSize;
+        this.responseSize = responseSize;
+    }
+
+    /**
+     * Constructs {@link StructuredLog} from {@link RequestContext} and {@link ResponseLog}.
+     * Can be used as {@link StructuredLogBuilder}.
+     */
+    protected StructuredLog(ResponseLog resLog) {
+        RequestLog reqLog = resLog.request();
+
+        responseTimeNanos = reqLog.endTimeNanos() - reqLog.startTimeNanos();
+        // Calculate the time of when received the request
+        timestampMillis = TimeUnit.NANOSECONDS.toMillis(
+                TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis()) - responseTimeNanos);
```

Comment:
Agreed, but I'm not sure if we should name it `startTimeMillis`.

It sounds like a method which provides the same thing to `startTimeNanos` just in different precision, when there's the semantic difference between `System.currentTimeMillis` and `System.nanoTime`.

Maybe `receiveRequestTimeMillis()` or something?